whitepapers/0011_accountant.md

```diff
@@ -0,0 +1,191 @@
+
+`global-accountant`keeps track of the tokens locked and minted on each connected blockchain.
+
+Once a Token Bridge message has reached a quorum of pre-observations (i.e. 13 guardians have submitted a pre-observation), it attempts to apply the transfer to the account balances for each chain.  If the transfer doesn't result in a negative account balance, it is allowed to proceed.
+
+## Detailed Design
+
+### Accountant Smart Contract
+
+#### Query methods for signature verification
+
+The global-accountant is a cosmwasm-based smart contract rather than a built-in native cosmos-sdk module.  It needs to expose query methods from the native wormhole module to cosmwasm (like getting the members of a guardian set).  Luckily, extending the contract interface is [fairly straightforward](https://docs.cosmwasm.com/docs/1.0/integration#extending-the-contract-interface).  The `x/wormhole` native module will expose the following query methods:
+
+```rust
+#[cw_serde]
+#[derive(QueryResponses)]
+pub enum WormholeQuery {
+    /// Verifies that `vaa` has been signed by a quorum of guardians from valid guardian set.
+    #[returns(Empty)]
+    VerifyVaa {
+        vaa: Binary,
+    },
+
+    /// Verifies that `data` has been signed by a guardian from `guardian_set_index` using a correct `prefix`.
+    #[returns(Empty)]
+    VerifyMessageSignature {
+        prefix: Binary,
+        data: Binary,
+        guardian_set_index: u32,
+        signature: Signature,
+    },
+
+    /// Returns the number of signatures necessary for quorum for the given guardian set index.
+    #[returns(u32)]
+    CalculateQuorum { guardian_set_index: u32 },
+}
+```
+
+#### Submitting observations to wormchain
+
+When a guardian observes a token transfer message from the registered tokenbridge on a particular chain, it submits the observation directly to the accountant.  The contract keeps track of pending token transfers and emits an event once it has received a quorum of observations.  If a quorum of guardians submit their observations within a single tendermint block then the token transfer only needs to wait for one round of tendermint consensus and one round of guardian consensus.
+
+When a guardian submits an observation to the contract, the contract will perform several checks:
+
+- Check that the signature is valid and from a guardian in the current guardian set.
+- If a transfer with the same `(emitter_chain, emitter_address, sequence)` tuple has already been committed, then check that the digest of the observation matches the digest of the committed transfer.  If they match return a `Committed` status; otherwise return an `Error` status and emit an `Error` event.
+- If the transfer has not been committed, mark the observation as having a signature from the guardian.
+- If the observation does not yet have a quorum of signatures, return a `Pending` status.
+- If the observation has a quorum of signatures, parse the tokenbridge payload and commit the transfer.
```

Comment:
replace observation with transfer

Reply:
Fixed in 8abbb46a902fec028dd9f41a834a4a30765a9588